mkosi/addon.py

```diff
@@ -0,0 +1,212 @@
+import contextlib
+import os
+import platform
+import shutil
+import sys
+import tempfile
+from pathlib import Path
+from typing import cast
+
+import mkosi.resources
+from mkosi.config import DocFormat, OutputFormat
+from mkosi.documentation import show_docs
+from mkosi.log import log_notice, log_setup
+from mkosi.run import find_binary, run, uncaught_exception_handler
+from mkosi.sandbox import __version__, umask
+from mkosi.tree import copy_tree
+from mkosi.types import PathString
+from mkosi.util import resource_path
+
+
+@uncaught_exception_handler()
+def main() -> None:
+    log_setup()
+
+    parser = argparse.ArgumentParser(
+        prog="mkosi-addon",
+        description="Build initrd/cmdline/ucode addon for the current system using mkosi",
+        allow_abbrev=False,
+        usage="mkosi-addon [options...]",
+    )
+
+    parser.add_argument(
+        "--kernel-version",
+        metavar="KERNEL_VERSION",
+        help="Kernel version string",
+        default=platform.uname().release,
+    )
+    parser.add_argument(
+        "-o",
+        "--output",
+        metavar="NAME",
+        help="Output name",
+        default="mkosi-local.addon.efi",
+    )
+    parser.add_argument(
+        "-O",
+        "--output-dir",
+        metavar="DIR",
+        help="Output directory",
+        default="",
+    )
+    parser.add_argument(
+        "--debug",
+        help="Turn on debugging output",
+        action="store_true",
+        default=False,
+    )
+    parser.add_argument(
+        "--debug-shell",
+        help="Spawn debug shell if a sandboxed command fails",
+        action="store_true",
+        default=False,
+    )
+    parser.add_argument(
+        "-D",
+        "--show-documentation",
+        help="Show the man page",
+        action="store_true",
+        default=False,
+    )
+    parser.add_argument(
+        "--version",
+        action="version",
+        version=f"mkosi {__version__}",
+    )
+
+    args = parser.parse_args()
+
+    if args.show_documentation:
+        with resource_path(mkosi.resources) as r:
+            show_docs("mkosi-addon", DocFormat.all(), resources=r)
+        return
+
+    with (
+        tempfile.TemporaryDirectory() as staging_dir,
+        tempfile.TemporaryDirectory() as sandbox_tree,
+    ):
+        cmdline: list[PathString] = [
+            "mkosi",
+            "--force",
+            "--directory", "",
+            f"--format={str(OutputFormat.addon)}",
+            "--output", args.output,
+            "--output-directory", staging_dir,
+            "--build-sources", "",
+            "--include=mkosi-addon",
+        ]  # fmt: skip
+
+        cmdline += [
+            "--extra-tree", f"/usr/lib/modules/{args.kernel_version}:/usr/lib/modules/{args.kernel_version}",
+            "--extra-tree=/usr/lib/firmware:/usr/lib/firmware",
+            "--remove-files=/usr/lib/firmware/*-ucode",
+            "--kernel-modules-exclude=.*",
+            "--kernel-modules-include=host",
```

Comment:
I don't think we should do this by default, let's leave it for users to enable in their mkosi-addon configuration.